File: C++/updateMatrix.cpp
/**
 * @File    : updateMatrix.cpp
 * @Brief   : 01 矩阵 动态规划  
 * @Link    : https://leetcode-cn.com/problems/01-matrix/
 * @Author  : Wei Li
 * @Date    : 2021-07-14
*/

#include <iostream>
#include <vector>
#include <algorithm>

class Solution
{
public:
    std::vector<std::vector<int>> updateMatrix(std::vector<std::vector<int>> &matrix)
    {
        if (matrix.empty())
            return {};
        int rows = matrix.size(), columns = matrix[0].size();
        std::vector<std::vector<int>> dp(rows, std::vector<int>(columns, INT_MAX - 1));
        // 从 左上 到 右下 的动态搜索
        for (int i = 0; i < rows; ++i)
        {
            for (int j = 0; j < columns; ++j)
            {
                if (matrix[i][j] == 0)
                {
                    dp[i][j] = 0;
                }
                else
                {
                    if (j > 0)
                    {
                        dp[i][j] = std::min(dp[i][j], dp[i][j - 1] + 1);
                    }
                    if (i > 0)
                    {
                        dp[i][j] = std::min(dp[i][j], dp[i - 1][j] + 1);
                    }
                }
            }
        }

        // 从 右下 到 左上 的动态搜索
        for (int i = rows - 1; i >= 0; --i)
        {
            for (int j = columns - 1; j >= 0; --j)
            {
                if (matrix[i][j] != 0)
                {
                    if (j < columns - 1)
                    {
                        dp[i][j] = std::min(dp[i][j], dp[i][j + 1] + 1);
                    }
                    if (i < rows - 1)
                    {
                        dp[i][j] = std::min(dp[i][j], dp[i + 1][j] + 1);
                    }
                }
            }
        }

        return dp;
    }
};

template <typename T>
void DisplayElement(const T &container)
{
    for (auto element = container.begin(); element != container.end(); ++element)
    {
        for (auto value = element->begin(); value != element->end(); ++value)
        {
            std::cout << *value << " ";
        }
        std::cout << std::endl;
    }
    std::cout << std::endl;
}

// -----------------------------
int main(int argc, char **argv)
{
    std::vector<std::vector<int>> matrix = { {0, 0, 0},
                                             {0, 1, 0},
                                             {1, 1, 1} };
    // std::vector<std::vector<int>> matrix = { {0, 0, 0},
    //                                          {0, 1, 0},
    //                                          {0, 0, 0} };

    Solution solution;
    auto update_mat = solution.updateMatrix(matrix);
    std::cout << "Before Update the Matrix: " << std::endl;
    DisplayElement(matrix);

    std::cout << "After Update the Matrix: " << std::endl;
    DisplayElement(update_mat);

    return 0;
}

File: Python/updateMatrix.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: updateMatrix.py
@Function: 01 矩阵 动态规划 
@Link: https://leetcode-cn.com/problems/01-matrix/
@Python Version: 3.8
@Author: Wei Li
@Date: 2021-07-14
"""

from typing import List


class Solution:
    def updateMatrix(self, matrix: List[List[int]]) -> List[List[int]]:
        m, n = len(matrix), len(matrix[0])
        # 初始化动态规划的数组，所有的距离值都设置为一个很大的数
        dist = [[10**9] * n for _ in range(m)]
        # 如果 (i, j) 的元素为 0，那么距离为 0
        for i in range(m):
            for j in range(n):
                if matrix[i][j] == 0:
                    dist[i][j] = 0
        # 只有 水平向左移动 和 竖直向上移动，注意动态规划的计算顺序
        for i in range(m):
            for j in range(n):
                if i - 1 >= 0:
                    dist[i][j] = min(dist[i][j], dist[i - 1][j] + 1)
                if j - 1 >= 0:
                    dist[i][j] = min(dist[i][j], dist[i][j - 1] + 1)
        # 只有 水平向右移动 和 竖直向下移动，注意动态规划的计算顺序
        for i in range(m - 1, -1, -1):
            for j in range(n - 1, -1, -1):
                if i + 1 < m:
                    dist[i][j] = min(dist[i][j], dist[i + 1][j] + 1)
                if j + 1 < n:
                    dist[i][j] = min(dist[i][j], dist[i][j + 1] + 1)
        return dist


# --------------------------
if __name__ == "__main__":
    matrix = [[0, 0, 0],
              [0, 1, 0],
              [1, 1, 1]]
    # matrix = [[0, 0, 0],
    #           [0, 1, 0],
    #           [0, 0, 0]]

    solution = Solution()
    update_mat = solution.updateMatrix(matrix)
    print(f"The matrix before update: \n {matrix}")
    print(f"\nThe matrix after update: \n {update_mat}")
